syntax = "proto3";

package hermes;

// Search service
service SearchService {
  // Search for documents
  rpc Search(SearchRequest) returns (SearchResponse);

  // Get document by ID
  rpc GetDocument(GetDocumentRequest) returns (GetDocumentResponse);

  // Get index info
  rpc GetIndexInfo(GetIndexInfoRequest) returns (GetIndexInfoResponse);
}

// Index service
service IndexService {
  // Create a new index (supports both structured schema and SDL string)
  rpc CreateIndex(CreateIndexRequest) returns (CreateIndexResponse);

  // Add documents to index (streaming)
  rpc IndexDocuments(stream IndexDocumentRequest) returns (IndexDocumentsResponse);

  // Add documents in batch
  rpc BatchIndexDocuments(BatchIndexDocumentsRequest) returns (BatchIndexDocumentsResponse);

  // Commit pending changes
  rpc Commit(CommitRequest) returns (CommitResponse);

  // Force merge segments
  rpc ForceMerge(ForceMergeRequest) returns (ForceMergeResponse);

  // Delete an index
  rpc DeleteIndex(DeleteIndexRequest) returns (DeleteIndexResponse);
}

// Query types
message Query {
  oneof query {
    TermQuery term = 1;
    BooleanQuery boolean = 2;
    BoostQuery boost = 3;
    AllQuery all = 4;
    SparseVectorQuery sparse_vector = 5;
    DenseVectorQuery dense_vector = 6;
  }
}

// How to combine scores for multi-valued documents
enum MultiValueCombiner {
  COMBINER_SUM = 0;  // Sum all scores (default)
  COMBINER_MAX = 1;  // Take maximum score
  COMBINER_AVG = 2;  // Average all scores
}

// Sparse vector query for similarity search
// Either provide (indices, values) directly, or provide text for server-side tokenization
message SparseVectorQuery {
  string field = 1;
  repeated uint32 indices = 2;   // Pre-computed token indices
  repeated float values = 3;     // Pre-computed token values
  string text = 4;               // Raw text (tokenized server-side if tokenizer configured)
  MultiValueCombiner combiner = 5;  // How to combine scores for multi-value fields
  float heap_factor = 6;         // Approximate search factor (1.0 = exact, 0.8 = ~20% faster)
}

// Dense vector query for similarity search
message DenseVectorQuery {
  string field = 1;
  repeated float vector = 2;
  uint32 nprobe = 3;           // Number of clusters to probe (for IVF indexes)
  uint32 rerank_factor = 4;    // Re-ranking factor (multiplied by k)
  MultiValueCombiner combiner = 5;  // How to combine scores for multi-value fields
}

message TermQuery {
  string field = 1;
  string term = 2;
}

message BooleanQuery {
  repeated Query must = 1;
  repeated Query should = 2;
  repeated Query must_not = 3;
}

message BoostQuery {
  Query query = 1;
  float boost = 2;
}

message AllQuery {}

// Search request/response
message SearchRequest {
  string index_name = 1;
  Query query = 2;
  uint32 limit = 3;
  uint32 offset = 4;
  repeated string fields_to_load = 5;
}

message SearchHit {
  uint32 doc_id = 1;
  float score = 2;
  map<string, FieldValue> fields = 3;
  repeated OrdinalScore ordinal_scores = 4;  // Per-ordinal scores for multi-value fields
}

// Score contribution from a specific ordinal in a multi-valued field
message OrdinalScore {
  uint32 ordinal = 1;  // Which value in the multi-valued field (0-indexed)
  float score = 2;     // Score contribution from this ordinal
}

message FieldValue {
  oneof value {
    string text = 1;
    uint64 u64 = 2;
    int64 i64 = 3;
    double f64 = 4;
    bytes bytes_value = 5;
    SparseVector sparse_vector = 6;
    DenseVector dense_vector = 7;
    string json_value = 8;  // JSON serialized as string
  }
}

// Sparse vector with term indices and weights
message SparseVector {
  repeated uint32 indices = 1;
  repeated float values = 2;
}

// Dense vector (float32 values)
message DenseVector {
  repeated float values = 1;
}

message SearchResponse {
  repeated SearchHit hits = 1;
  uint32 total_hits = 2;
  uint64 took_ms = 3;
}

// Get document request/response
message GetDocumentRequest {
  string index_name = 1;
  uint32 doc_id = 2;
}

message GetDocumentResponse {
  map<string, FieldValue> fields = 1;
}

// Index info request/response
message GetIndexInfoRequest {
  string index_name = 1;
}

message GetIndexInfoResponse {
  string index_name = 1;
  uint32 num_docs = 2;
  uint32 num_segments = 3;
  // Schema in SDL format
  string schema = 4;
}

// Create index request/response
message CreateIndexRequest {
  string index_name = 1;
  // Schema definition
  string schema = 2;
}

message CreateIndexResponse {
  bool success = 1;
}

// Field entry for multi-value field support
message FieldEntry {
  string name = 1;
  FieldValue value = 2;
}

// Named document for batch indexing
// Uses repeated FieldEntry to support multi-value fields (same name, multiple values)
message NamedDocument {
  repeated FieldEntry fields = 1;
}

// Batch index documents request
message BatchIndexDocumentsRequest {
  string index_name = 1;
  repeated NamedDocument documents = 2;
}

message BatchIndexDocumentsResponse {
  uint32 indexed_count = 1;
  uint32 error_count = 2;
}

// Index document request/response
message IndexDocumentRequest {
  string index_name = 1;
  repeated FieldEntry fields = 2;
}

message IndexDocumentsResponse {
  uint32 indexed_count = 1;
}

// Commit request/response
message CommitRequest {
  string index_name = 1;
}

message CommitResponse {
  bool success = 1;
  uint32 num_docs = 2;
}

// Force merge request/response
message ForceMergeRequest {
  string index_name = 1;
}

message ForceMergeResponse {
  bool success = 1;
  uint32 num_segments = 2;
}

// Delete index request/response
message DeleteIndexRequest {
  string index_name = 1;
}

message DeleteIndexResponse {
  bool success = 1;
}
